File: lab_2/src/generators/generator.cpp
// lab_2/src/generators/generator.cpp
// Генератор псевдослучайной последовательности на C++
// Использует std::mt19937 (Mersenne Twister) — стандарт C++11

#include <iostream>
#include <fstream>
#include <random>
#include <string>
#include <filesystem>

namespace fs = std::filesystem;

/**
 * Генерирует 128-битную бинарную последовательность
 * и сохраняет в файл с ОТНОСИТЕЛЬНЫМ путём
 */
int main() {
    // Инициализация генератора
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> distrib(0, 1);
    
    // Генерация 128 бит
    std::string sequence;
    sequence.reserve(128);
    
    for (int i = 0; i < 128; ++i) {
        sequence += std::to_string(distrib(gen));
    }
    
    // ✅ ОТНОСИТЕЛЬНЫЙ ПУТЬ: работает из любой директории
    fs::path output_path = fs::path("../../data/sequences/cpp_sequence.txt");
    fs::create_directories(output_path.parent_path());
    
    std::ofstream file(output_path);
    if (file.is_open()) {
        file << sequence;
        file.close();
        std::cout << "[C++] Generated: " << sequence << "\n";
        std::cout << "[C++] Saved: " << output_path << "\n";
    } else {
        std::cerr << "[C++] Error: Cannot open " << output_path << "\n";
        return 1;
    }
    
    return 0;
}

File: lab_2/lab_2/src/generators/Generator.java
// lab_2/src/generators/Generator.java
// Генератор псевдослучайной последовательности на Java
// Использует java.util.Random

import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Random;

public class Generator {
    
    private static final int SEQUENCE_LENGTH = 128;
    
    public static void main(String[] args) {
        try {
            String sequence = generateBinarySequence(SEQUENCE_LENGTH);
            
            // ✅ ОТНОСИТЕЛЬНЫЙ ПУТЬ через Path
            Path outputPath = Paths.get("../../data/sequences/java_sequence.txt")
                                   .normalize()
                                   .toAbsolutePath();
            Files.createDirectories(outputPath.getParent());
            
            saveToFile(sequence, outputPath.toString());
            
            System.out.println("[Java] Generated: " + sequence);
            System.out.println("[Java] Saved: " + outputPath);
            
        } catch (IOException e) {
            System.err.println("[Java] Error: " + e.getMessage());
            System.exit(1);
        }
    }
    
    /**
     * Генерирует бинарную последовательность заданной длины
     */
    private static String generateBinarySequence(int length) {
        Random random = new Random(System.currentTimeMillis());
        StringBuilder sb = new StringBuilder(length);
        
        for (int i = 0; i < length; i++) {
            sb.append(random.nextInt(2));
        }
        return sb.toString();
    }
    
    /**
     * Сохраняет данные в файл
     */
    private static void saveToFile(String data, String filepath) throws IOException {
        try (FileWriter writer = new FileWriter(filepath)) {
            writer.write(data);
        }
    }
}